File: backend/mcp-code-intelligence-nodejs/src/analyzers/entry-points/RouteResolver.ts
/**
 * KSA-162: Route Resolver — Resolves full routes from controller prefix + method path.
 */

export class RouteResolver {
  /** Resolve full route: controller prefix + method path. */
  resolve(controllerPrefix: string | null, methodPath: string): string {
    const prefix = this.normalizePath(controllerPrefix ?? '');
    const path = this.normalizePath(methodPath);

    if (!prefix) return path || '/';
    if (!path || path === '/') return prefix;
    return `${prefix}${path}`;
  }

  /** Normalize route params: :id, {id}, <int:id> → {id}. */
  normalizeParams(path: string): string {
    return path
      // Express-style :param → {param}
      .replace(/:([a-zA-Z_][a-zA-Z0-9_]*)/g, '{$1}')
      // Flask-style <type:param> → {param}
      .replace(/<(?:[a-z]+:)?([a-zA-Z_][a-zA-Z0-9_]*)>/g, '{$1}')
      // Already normalized {param} stays as-is
      ;
  }

  /** Extract route path from a decorator/annotation argument string. */
  extractPathFromArg(arg: string): string {
    // Remove quotes
    let path = arg.replace(/^['"`]|['"`]$/g, '').trim();
    // Normalize params
    path = this.normalizeParams(path);
    return path;
  }

  private normalizePath(path: string): string {
    if (!path) return '';
    // Ensure starts with /
    if (!path.startsWith('/')) path = '/' + path;
    // Remove trailing slash (except root)
    if (path.length > 1 && path.endsWith('/')) path = path.slice(0, -1);
    return path;
  }
}

File: backend/mcp-code-intelligence-kotlin/bin/main/com/codeintel/analyzers/entrypoints/RouteResolver.kt
/**
 * KSA-162: Route Resolver — Resolves full routes from prefix + method path.
 */
package com.codeintel.analyzers.entrypoints

class RouteResolver {

    fun resolve(controllerPrefix: String?, methodPath: String): String {
        val prefix = normalizePath(controllerPrefix ?: "")
        val path = normalizePath(methodPath)
        if (prefix.isEmpty()) return path.ifEmpty { "/" }
        if (path.isEmpty() || path == "/") return prefix
        return "$prefix$path"
    }

    fun normalizeParams(path: String): String =
        path.replace(Regex("<(?:[a-z]+:)?([a-zA-Z_][a-zA-Z0-9_]*)>"), "{$1}")
            .replace(Regex(":([a-zA-Z_][a-zA-Z0-9_]*)"), "{$1}")

    fun extractPathFromArg(arg: String): String {
        val path = arg.trim().removeSurrounding("\"").removeSurrounding("'").removeSurrounding("`")
        return normalizeParams(path)
    }

    private fun normalizePath(path: String): String {
        if (path.isEmpty()) return ""
        var p = path
        if (!p.startsWith("/")) p = "/$p"
        if (p.length > 1 && p.endsWith("/")) p = p.dropLast(1)
        return p
    }
}
